extensions-core/druid-basic-security/src/main/java/io/druid/security/basic/authorization/db/updater/CoordinatorBasicAuthorizerMetadataStorageUpdater.java

```diff
@@ -0,0 +1,761 @@
+import io.druid.metadata.MetadataCASUpdate;
+import io.druid.metadata.MetadataStorageConnector;
+import io.druid.metadata.MetadataStorageTablesConfig;
+import io.druid.security.basic.BasicSecurityDBResourceException;
+import io.druid.security.basic.BasicAuthCommonCacheConfig;
+import io.druid.security.basic.authorization.BasicRoleBasedAuthorizer;
+import io.druid.security.basic.authorization.db.cache.BasicAuthorizerCacheNotifier;
+import io.druid.security.basic.authorization.entity.BasicAuthorizerPermission;
+import io.druid.security.basic.authorization.entity.BasicAuthorizerRole;
+import io.druid.security.basic.authorization.entity.BasicAuthorizerUser;
+import io.druid.security.basic.authorization.entity.UserAndRoleMap;
+import io.druid.server.security.Action;
+import io.druid.server.security.Authorizer;
+import io.druid.server.security.AuthorizerMapper;
+import io.druid.server.security.Resource;
+import io.druid.server.security.ResourceAction;
+import io.druid.server.security.ResourceType;
+import org.joda.time.Duration;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+@ManageLifecycle
+public class CoordinatorBasicAuthorizerMetadataStorageUpdater implements BasicAuthorizerMetadataStorageUpdater
+{
+  private static final EmittingLogger LOG =
+      new EmittingLogger(CoordinatorBasicAuthorizerMetadataStorageUpdater.class);
+
+  public static final List<ResourceAction> SUPERUSER_PERMISSIONS = makeSuperUserPermissions();
+
+  public static final String USERS = "users";
+  public static final String ROLES = "roles";
+  public static final TypeReference USER_MAP_TYPE_REFERENCE = new TypeReference<Map<String, BasicAuthorizerUser>>()
+  {
+  };
+  public static final TypeReference ROLE_MAP_TYPE_REFERENCE = new TypeReference<Map<String, BasicAuthorizerRole>>()
+  {
+  };
+  public static final TypeReference USER_AND_ROLE_MAP_TYPE_REFERENCE = new TypeReference<UserAndRoleMap>()
+  {
+  };
+
+  public static final String DEFAULT_ADMIN_NAME = "admin";
+  public static final String DEFAULT_INTERNAL_SYSTEM_NAME = "druid_system";
+  
+  private final AuthorizerMapper authorizerMapper;
+  private final MetadataStorageConnector connector;
+  private final MetadataStorageTablesConfig connectorConfig;
+  private final BasicAuthorizerCacheNotifier cacheNotifier;
+  private final BasicAuthCommonCacheConfig commonCacheConfig;
+  private final ObjectMapper objectMapper;
+  private final int numRetries = 5;
+
+  private final Map<String, Map<String, BasicAuthorizerUser>> cachedUserMaps;
+  private final Map<String, byte[]> cachedSerializedUserMaps;
+
+  private final Map<String, Map<String, BasicAuthorizerRole>> cachedRoleMaps;
+  private final Map<String, byte[]> cachedSerializedRoleMaps;
```

Comment:
Looks that we can delegate thread-safety to ConcurrentHashMap if cachedUserMaps and cachedRoleMaps are changed to something like ConcurrentHashMap<String, Pair<Map<String, BasicAuthorizerUser(/Role)>, byte[]>>. Also suggest to create a class to represent Pair<Map<String, BasicAuthorizerUser(/Role)>, byte[]> for better readability.

Reply:
Added *Bundle class for the combined map and serialized version, changed to ConcurrentHashMaps